// prettier-plugin-solidity/dist/slang-nodes/AssignmentExpression.js
import { NonterminalKind } from '@nomicfoundation/slang/cst';
import { extractVariant } from '../slang-utils/extract-variant.js';
import { printAssignmentRightSide } from '../slang-printers/print-assignment-right-side.js';
import { SlangNode } from './SlangNode.js';
import { Expression } from './Expression.js';
export class AssignmentExpression extends SlangNode {
    constructor(ast, collected, options) {
        super(ast, collected);
        this.kind = NonterminalKind.AssignmentExpression;
        this.leftOperand = extractVariant(new Expression(ast.leftOperand, collected, options));
        this.operator = ast.operator.unparse();
        this.rightOperand = extractVariant(new Expression(ast.rightOperand, collected, options));
        this.updateMetadata(this.leftOperand, this.rightOperand);
    }
    print(path, print) {
        return [
            path.call(print, 'leftOperand'),
            ` ${this.operator}`,
            printAssignmentRightSide(path.call(print, 'rightOperand'), this.rightOperand)
        ];
    }
}
