// prettier-plugin-solidity/dist/slang-nodes/FallbackFunctionDefinition.js
import { NonterminalKind } from '@nomicfoundation/slang/cst';
import { printFunctionWithBody } from '../slang-printers/print-function.js';
import { extractVariant } from '../slang-utils/extract-variant.js';
import { SlangNode } from './SlangNode.js';
import { ParametersDeclaration } from './ParametersDeclaration.js';
import { FallbackFunctionAttributes } from './FallbackFunctionAttributes.js';
import { ReturnsDeclaration } from './ReturnsDeclaration.js';
import { FunctionBody } from './FunctionBody.js';
export class FallbackFunctionDefinition extends SlangNode {
    constructor(ast, collected, options) {
        super(ast, collected);
        this.kind = NonterminalKind.FallbackFunctionDefinition;
        this.parameters = new ParametersDeclaration(ast.parameters, collected, options);
        this.attributes = new FallbackFunctionAttributes(ast.attributes, collected, options);
        if (ast.returns) {
            this.returns = new ReturnsDeclaration(ast.returns, collected, options);
        }
        this.body = extractVariant(new FunctionBody(ast.body, collected, options));
        this.updateMetadata(this.parameters, this.attributes, this.returns, this.body);
        for (const attribute of this.attributes.items) {
            if (attribute.kind === NonterminalKind.ModifierInvocation) {
                attribute.cleanModifierInvocationArguments();
            }
        }
    }
    print(path, print) {
        return printFunctionWithBody('fallback', this, path, print);
    }
}
