// prettier-plugin-solidity/dist/slang-nodes/FunctionDefinition.js
import { NonterminalKind } from '@nomicfoundation/slang/cst';
import { satisfies } from 'semver';
import { printFunctionWithBody } from '../slang-printers/print-function.js';
import { extractVariant } from '../slang-utils/extract-variant.js';
import { SlangNode } from './SlangNode.js';
import { FunctionName } from './FunctionName.js';
import { ParametersDeclaration } from './ParametersDeclaration.js';
import { FunctionAttributes } from './FunctionAttributes.js';
import { ReturnsDeclaration } from './ReturnsDeclaration.js';
import { FunctionBody } from './FunctionBody.js';
export class FunctionDefinition extends SlangNode {
    constructor(ast, collected, options) {
        super(ast, collected);
        this.kind = NonterminalKind.FunctionDefinition;
        this.name = extractVariant(new FunctionName(ast.name, collected));
        this.parameters = new ParametersDeclaration(ast.parameters, collected, options);
        this.attributes = new FunctionAttributes(ast.attributes, collected, options);
        if (ast.returns) {
            this.returns = new ReturnsDeclaration(ast.returns, collected, options);
        }
        this.body = extractVariant(new FunctionBody(ast.body, collected, options));
        this.updateMetadata(this.name, this.parameters, this.attributes, this.returns, this.body);
        // Older versions of Solidity defined a constructor as a function having
        // the same name as the contract.
        // So we delegate to the parents the responsibility of cleaning the
        // arguments of modifier invocations.
        if (satisfies(options.compiler, '>=0.5.0')) {
            this.cleanModifierInvocationArguments();
        }
    }
    cleanModifierInvocationArguments() {
        for (const attribute of this.attributes.items) {
            if (attribute.kind === NonterminalKind.ModifierInvocation) {
                attribute.cleanModifierInvocationArguments();
            }
        }
    }
    print(path, print) {
        return printFunctionWithBody(['function ', path.call(print, 'name')], this, path, print);
    }
}
