// prettier-plugin-solidity/dist/slang-nodes/FunctionType.js
import { NonterminalKind } from '@nomicfoundation/slang/cst';
import { printFunction } from '../slang-printers/print-function.js';
import { SlangNode } from './SlangNode.js';
import { ParametersDeclaration } from './ParametersDeclaration.js';
import { FunctionTypeAttributes } from './FunctionTypeAttributes.js';
import { ReturnsDeclaration } from './ReturnsDeclaration.js';
export class FunctionType extends SlangNode {
    constructor(ast, collected, options) {
        super(ast, collected);
        this.kind = NonterminalKind.FunctionType;
        this.parameters = new ParametersDeclaration(ast.parameters, collected, options);
        this.attributes = new FunctionTypeAttributes(ast.attributes, collected);
        if (ast.returns) {
            this.returns = new ReturnsDeclaration(ast.returns, collected, options);
        }
        this.updateMetadata(this.parameters, this.attributes, this.returns);
    }
    print(path, print) {
        return printFunction('function', this, path, print);
    }
}
