// prettier-plugin-solidity/dist/slang-nodes/LibraryDefinition.js
import { NonterminalKind } from '@nomicfoundation/slang/cst';
import { doc } from 'prettier';
import { satisfies } from 'semver';
import { SlangNode } from './SlangNode.js';
import { TerminalNode } from './TerminalNode.js';
import { LibraryMembers } from './LibraryMembers.js';
const { group, line } = doc.builders;
export class LibraryDefinition extends SlangNode {
    constructor(ast, collected, options) {
        super(ast, collected);
        this.kind = NonterminalKind.LibraryDefinition;
        this.name = new TerminalNode(ast.name, collected);
        this.members = new LibraryMembers(ast.members, collected, options);
        this.updateMetadata(this.members);
        // Older versions of Solidity defined a constructor as a function having
        // the same name as the contract.
        // So we delegate to the parents the responsibility of cleaning the
        // arguments of modifier invocations.
        if (!satisfies(options.compiler, '>=0.5.0')) {
            for (const member of this.members.items) {
                if (member.kind === NonterminalKind.FunctionDefinition) {
                    member.cleanModifierInvocationArguments();
                }
            }
        }
    }
    print(path, print) {
        return [
            group(['library ', path.call(print, 'name'), line, '{']),
            path.call(print, 'members'),
            '}'
        ];
    }
}
