// prettier-plugin-solidity/dist/slang-nodes/StateVariableDefinition.js
import { NonterminalKind } from '@nomicfoundation/slang/cst';
import { doc } from 'prettier';
import { printGroupAndIndentIfBreakPair } from '../slang-printers/print-group-and-indent-if-break-pair.js';
import { extractVariant } from '../slang-utils/extract-variant.js';
import { SlangNode } from './SlangNode.js';
import { TypeName } from './TypeName.js';
import { StateVariableAttributes } from './StateVariableAttributes.js';
import { TerminalNode } from './TerminalNode.js';
import { StateVariableDefinitionValue } from './StateVariableDefinitionValue.js';
const { indent } = doc.builders;
export class StateVariableDefinition extends SlangNode {
    constructor(ast, collected, options) {
        super(ast, collected);
        this.kind = NonterminalKind.StateVariableDefinition;
        this.typeName = extractVariant(new TypeName(ast.typeName, collected, options));
        this.attributes = new StateVariableAttributes(ast.attributes, collected);
        this.name = new TerminalNode(ast.name, collected);
        if (ast.value) {
            this.value = new StateVariableDefinitionValue(ast.value, collected, options);
        }
        this.updateMetadata(this.typeName, this.attributes, this.value);
    }
    print(path, print) {
        return printGroupAndIndentIfBreakPair([
            path.call(print, 'typeName'),
            indent(path.call(print, 'attributes')),
            ' ',
            path.call(print, 'name')
        ], [path.call(print, 'value'), ';']);
    }
}
