// prettier-plugin-solidity/dist/slang-nodes/TypedTupleMember.js
import { NonterminalKind } from '@nomicfoundation/slang/cst';
import { joinExisting } from '../slang-utils/join-existing.js';
import { extractVariant } from '../slang-utils/extract-variant.js';
import { SlangNode } from './SlangNode.js';
import { TypeName } from './TypeName.js';
import { StorageLocation } from './StorageLocation.js';
import { TerminalNode } from './TerminalNode.js';
export class TypedTupleMember extends SlangNode {
    constructor(ast, collected, options) {
        super(ast, collected);
        this.kind = NonterminalKind.TypedTupleMember;
        this.typeName = extractVariant(new TypeName(ast.typeName, collected, options));
        if (ast.storageLocation) {
            this.storageLocation = new StorageLocation(ast.storageLocation, collected);
        }
        this.name = new TerminalNode(ast.name, collected);
        this.updateMetadata(this.typeName, this.storageLocation);
    }
    print(path, print) {
        return joinExisting(' ', [
            path.call(print, 'typeName'),
            path.call(print, 'storageLocation'),
            path.call(print, 'name')
        ]);
    }
}
