// prettier-plugin-solidity/dist/slang-nodes/YulFunctionDefinition.js
import { NonterminalKind } from '@nomicfoundation/slang/cst';
import { SlangNode } from './SlangNode.js';
import { TerminalNode } from './TerminalNode.js';
import { YulParametersDeclaration } from './YulParametersDeclaration.js';
import { YulReturnsDeclaration } from './YulReturnsDeclaration.js';
import { YulBlock } from './YulBlock.js';
export class YulFunctionDefinition extends SlangNode {
    constructor(ast, collected, options) {
        super(ast, collected);
        this.kind = NonterminalKind.YulFunctionDefinition;
        this.name = new TerminalNode(ast.name, collected);
        this.parameters = new YulParametersDeclaration(ast.parameters, collected);
        if (ast.returns) {
            this.returns = new YulReturnsDeclaration(ast.returns, collected);
        }
        this.body = new YulBlock(ast.body, collected, options);
        this.updateMetadata(this.parameters, this.returns, this.body);
    }
    print(path, print) {
        return [
            'function ',
            path.call(print, 'name'),
            path.call(print, 'parameters'),
            path.call(print, 'returns') || ' ',
            path.call(print, 'body')
        ];
    }
}
